src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs

```diff
@@ -0,0 +1,315 @@
+
+                        // Copy chars from charsNotProcessedFromPreviousChunk to the combined array
+                        charsNotProcessedFromPreviousChunk.CopyTo(combinedArray);
+
+                        // Copy chars from chunk to the combined array, starting from the end of charsNotProcessedFromPreviousChunk
+                        chunk.CopyTo(combinedArray.AsMemory().Slice(charsNotProcessedFromPreviousChunk.Length));
+
+                        // Write the chunk depending on the firstIndexToEscape
+                        WriteChunk(combinedArray, totalLength, isFinalBlock, firstIndexToEscape);
+
+                        if (combinedArray != null)
+                        {
+                            ArrayPool<char>.Shared.Return(combinedArray);
+                        }
+                    }
+                    else
+                    {
+                        // Write the chunk directly depending on the firstIndexToEscape
+                        WriteChunk(chunk.Span, chunk.Length, isFinalBlock, firstIndexToEscape);
+                    }
+
+                    // Flush the writer if the buffer threshold is reached
+                    await this.jsonWriter.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+                }
+            }
+
+            /// <summary>
+            /// Writes a chunk of characters to the JSON writer, either escaped or directly, depending on the presence of characters that need escaping.
+            /// </summary>
+            /// <param name="chunk">The read-only span containing the chunk of characters to be written.</param>
+            /// <param name="chunkLength">The length of the read-only span containing the chunk of characters to be written.</param>
+            /// <param name="isFinalBlock">A boolean value indicating whether the current chunk is the final block of characters to be written.</param>
+            /// <param name="firstIndexToEscape">The index of the first character in the chunk that requires escaping, or -1 if no characters need escaping.</param>
+            private void WriteChunk(ReadOnlySpan<char> chunk, int chunkLength, bool isFinalBlock, int firstIndexToEscape)
+            {
+                if (firstIndexToEscape != -1)
+                { 
+                    this.jsonWriter.WriteEscapedStringChunk(chunk.Slice(0, chunkLength), firstIndexToEscape, isFinalBlock, out charsNotWrittenFromPreviousChunk);
+
+                    if (charsNotWrittenFromPreviousChunk > 0)
+                    {
+                        if (this.buffer == null)
+                        {
+                            this.buffer = new char[chunkSize];
+                        }
+
+                        // Update the buffer with unprocessed chars from the current chunk.
+                        chunk.Slice(chunkLength - charsNotWrittenFromPreviousChunk).CopyTo(this.buffer.AsSpan(bufferPosition));

```

Comment:
I would suggest that this buffer only store the left-over bytes from the most recent write. That way, the bufferPosition will always be 0 and we only need to keep track of the length of the contents in the buffer. Also, this means we can keep the buffer really small.

Reply:
This is what I had done before and you suggested we do it this way.

Reply:
Updated